scripts/data_utils.py

```diff
@@ -36,11 +38,202 @@
+        newTokens = self.GPT2_TOKENIZER.decode(
+            self.GPT2_TOKENIZER.encode(tokens, allowed_special="all")[:numofTokens]
+        )
+        return newTokens
+
+TOKEN_ESTIMATOR = TokenEstimator()
+
+class PdfTextSplitter(TextSplitter):
+    def __init__(self, length_function: Callable[[str], int] =TOKEN_ESTIMATOR.estimate_tokens, separator: str = "\n\n", **kwargs: Any):
+        """Create a new TextSplitter for htmls from extracted pdfs."""
+        super().__init__(**kwargs)
+        self._table_tags = HTML_TABLE_TAGS
+        self._separators = separator or ["\n\n", "\n", " ", ""]
+        self._length_function = length_function
+
+    def extract_caption(self, s, type):
+        separator = self._separators[-1]
+        for _s in self._separators:
+            if _s == "":
+                separator = _s
+                break
+            if _s in s:
+                separator = _s
+                break
+        # Now that we have the separator, split the text
+        if separator:
+            lines = s.split(separator)
+        else:
+            lines = list(s)
+
+        caption = ""
+        if type == "prefix": #find the last heading and the last line before the table
+            if len(s.split(f"<{PDF_HEADERS['title']}>"))>1:
+               caption +=  s.split(f"<{PDF_HEADERS['title']}>")[-1].split(f"</{PDF_HEADERS['title']}>")[0]
+            if len(s.split(f"<{PDF_HEADERS['sectionHeading']}>"))>1:
+               caption +=  s.split(f"<{PDF_HEADERS['sectionHeading']}>")[-1].split(f"</{PDF_HEADERS['sectionHeading']}>")[0]
+            caption += "\n"+ lines[-1]
+        else: # find the first line after the table
+            caption += lines[0]
+        return caption
+    
+    def split_text(self, text: str) -> List[str]:
+        start_tag = self._table_tags["table_open"]
+        end_tag = self._table_tags["table_close"]
+        splits = text.split(start_tag)
+        
+        chunks = [splits[0]] # the first split is before the first table tag so it is regular text
+        is_table = [0] # to keep track of regular (0) and those containing tables (1)
+        table_caption_prefix = self.extract_caption(splits[0], "prefix")
+        for part in splits[1:]:
+            table, rest = part.split(end_tag)
+            table_caption_suffix = self.extract_caption(rest, "suffix")
+            table = start_tag + table + end_tag 
+            minitables = self.chunk_table(table, "\n".join([table_caption_prefix, table_caption_suffix]))
+            chunks.extend(minitables)
+            is_table.extend([1]*len(minitables))
+            
+            chunks.append(rest)
+            is_table.append(0)
+            table_caption_prefix = self.extract_caption(rest, "prefix")
+
+        self.final_chunks = []
+        print(f"Here are the chunk sizes before chunking regular text: {[self._length_function(x) for x in chunks]}")
+        print(f"Number of chunks before chunking regular text: {len(chunks)}")

```

Comment:
Do these print for every document? Probably can remove so the output doesn't get cluttered